src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,685 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
+    /// underlying data slice. Otherwise this method
+    /// will panic.
+    pub fn set_value<T: Aarch64RegisterData<N>, const N: usize>(&mut self, value: T) {
+        self.data.copy_from_slice(&value.to_bytes())
+    }
+}
+
+/// Old definition of a struct describing a aarch64 register.
+/// This type is only used to have a backward compatibility
+/// with old snapshot versions and should not be used anywhere
+/// else.
+#[derive(Debug, Clone, Copy, Versionize)]
+pub struct Aarch64RegisterOld {
+    /// ID of the register.
+    pub id: u64,
+    /// Register data.
+    pub data: u128,
+}
+
+impl Aarch64RegisterOld {
+    /// Converts `Aarch64RegisterOld`
+    /// into `Aarch64RegisterRef`.
+    pub fn as_reg_ref(&self) -> Aarch64RegisterRef {
+        // # Safety:
+        // `self.data` is a valid memory and slice size is valid for this type.
+        let data_ref = unsafe {
+            std::slice::from_raw_parts(
+                (&self.data as *const u128).cast::<u8>(),
+                std::mem::size_of::<u128>(),
+            )
+        };
+        match RegSize::from(reg_size(self.id)) {
+            RegSize::U32 => {
+                Aarch64RegisterRef::new(self.id, &data_ref[..std::mem::size_of::<u32>()])
+            }
+            RegSize::U64 => {
+                Aarch64RegisterRef::new(self.id, &data_ref[..std::mem::size_of::<u64>()])
+            }
+            RegSize::U128 => Aarch64RegisterRef::new(self.id, data_ref),
+            _ => unreachable!("Only 32, 64 and 128 bit wide registers are supported"),
+        }
+    }
+
+    /// Converts `Aarch64RegisterRef` into `Aarch64RegisterOld`.
+    pub fn from_reg_ref(reg_ref: Aarch64RegisterRef) -> versionize::VersionizeResult<Self> {
+        let reg = match reg_ref.size() {
+            RegSize::U32 => Self {
+                id: reg_ref.id,
+                data: u128::from(reg_ref.value::<u32, 4>()),
+            },
+            RegSize::U64 => Self {
+                id: reg_ref.id,
+                data: u128::from(reg_ref.value::<u64, 8>()),
+            },
+            RegSize::U128 => Self {
+                id: reg_ref.id,
+                data: reg_ref.value::<u128, 16>(),
+            },
+            _ => {
+                return Err(VersionizeError::Serialize(
+                    "Only 32, 64 and 128 bit wide registers are supported".to_string(),
+                ))
+            }
+        };
+        Ok(reg)
+    }
+}
+
+/// Trait for data types that can represent aarch64
+/// register data.
+pub trait Aarch64RegisterData<const N: usize> {
+    /// Create data type from slice
+    fn from_slice(slice: &[u8]) -> Self;
+    /// Convert data type to array of bytes
+    fn to_bytes(&self) -> [u8; N];
+}
+
+macro_rules! reg_data {
+    ($t:ty, $bytes: expr) => {
+        impl Aarch64RegisterData<$bytes> for $t {
+            fn from_slice(slice: &[u8]) -> Self {
+                let mut bytes = [0_u8; $bytes];
+                bytes.clone_from_slice(slice);
```

Comment:
can we use copy_from_slice here instead? copy* uses `memcpy` while clone* iterates the slice if the optimizer has a bad day [[1]]

[1]: https://github.com/rust-lang/rust/blob/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/slice/mod.rs#L2930-L2938

Reply:
Nice catch. I thought I used copy here. (even though for slices it would have used `memcopy` anyway)